src/net/peer/connect/connect.rs

```diff
@@ -81,170 +83,161 @@ quick_error! {
-                            None => Err(SingleConnectionError::ConnectionDropped),
-                            Some(HandshakeMessage::ChooseConnection) => {
-                                peer::from_handshaken_socket(
-                                    &handle,
-                                    socket,
-                                    their_uid,
-                                    CrustUser::Node,
-                                ).map_err(SingleConnectionError::Io)
-                            }
-                            Some(_msg) => Err(SingleConnectionError::UnexpectedMessage),
-                        })
-                })
-                .buffer_unordered(128)
-                .until(timeout.infallible())
-                .first_ok()
-                .map_err(ConnectError::AllConnectionsFailed)
-                .into_boxed()
-        };
+/// Executes handshake process for the given connections.
+fn handshake_outgoing_connections<UID: Uid, S>(
+    evloop_handle: &Handle,
+    connections: S,
+    our_connect_request: ConnectRequest<UID>,
+    their_id: UID,
+    name_hash: NameHash,
+) -> BoxStream<(Socket<HandshakeMessage<UID>>, UID), SingleConnectionError>
+where
+    S: Stream<Item = TcpStream, Error = SingleConnectionError> + 'static,
+{
+    let handle_copy = evloop_handle.clone();
+    connections
+        .map(move |stream| {
+            let peer_addr = unwrap!(stream.peer_addr());
+            Socket::wrap_tcp(&handle_copy, stream, peer_addr)
+        })
+        .and_then(move |socket| {
+            socket
+                .send((0, HandshakeMessage::Connect(our_connect_request.clone())))
+                .map_err(SingleConnectionError::Socket)
+        })
+        .and_then(move |socket| {
+            socket.into_future().map_err(|(err, _socket)| {
+                SingleConnectionError::Socket(err)
+            })
+        })
+        .and_then(move |(msg_opt, socket)| match msg_opt {
+            None => Err(SingleConnectionError::ConnectionDropped),
+            Some(HandshakeMessage::Connect(connect_request)) => {
+                validate_connect_request(their_id, name_hash, &connect_request)?;
+                Ok((socket, connect_request.uid))
+            }
+            Some(_msg) => Err(SingleConnectionError::UnexpectedMessage),
+        })
+        .into_boxed()
+}
 
-        Ok(chosen_peer)
-    };
+/// Sends connection info to "rendezvous connect" task and waits for connection.
+///
+/// If given p2p connection info is `None`, returns empty future.
+fn connect_p2p(
+    rendezvous_channel: UnboundedBiChannel<Bytes>,
+    conn_rx: Option<oneshot::Receiver<Result<TcpStream, SingleConnectionError>>>,
+    p2p_conn_info: Option<Vec<u8>>,
+) -> BoxFuture<TcpStream, SingleConnectionError> {
+    if p2p_conn_info.is_none() || conn_rx.is_none() {
+        return future::empty().into_boxed();
+    }
 
-    future::result(try()).flatten().into_boxed()
+    let conn_info = Bytes::from(unwrap!(p2p_conn_info));
```

Comment:
You don't need to use an `unwrap!` here. Instead of using `p2p_conn_info.is_none()` and then unwrapping, use a match statement to get the connection info and then use it directly.

More generally, never use a panicking dynamic check like `unwrap!` if you can get the same guarantee from the type system.

eg:

```rust
// BAD
if foo.is_none() {
    return;
}
let bar = unwrap!(foo);

// GOOD
let bar = match foo {
    Some(bar) => bar,
    None => return,
};
```

With the `BAD` code above, you may think you "know" that unwrapping is safe because foo is always `Some(..)`, but don't rely on knowing things when you can use the type checker to enforce them. Things that we "know" work can still break if we make a mistake, or if the code gets refactored and a bug gets introduced. Things that are checked by the type systems are guaranteed to work or will break at compile time if someone breaks the code. Also, using correct types makes code more readable - if you use an `unwrap` I have to look around and figure out why you're so sure the `Option` must contain a `Some`.